### ISSUE TITLE
performance issues on gpu box

### ISSUE THREAD

**Issue by fmorbinitc** [deeplearning4j/deeplearning4j#6718]
two machines:
cpu box is a apple laptop with 8 cores
gpu box is a p2.xlarge aws box which has 4 cpu cores and a tesla k80 gpu

reporting PerformanceListener for first 4 iterations, batch size always equal to 1000.

no particular mem configuration for the jvm on all machines.

**gpu machine (using gpu for both nd4j and cudnn, gc setup using: Nd4j.getMemoryManager().setOccasionalGcFrequency(10000))**

81148 [main] INFO  o.d.o.l.PerformanceListener - ETL: 15068 ms; iteration 0; iteration time: 30044 ms; samples/sec: 33.285; batches/sec: 0.033; score: 18.329551218376967;
98099 [main] INFO  o.d.o.l.PerformanceListener - ETL: 11029 ms; iteration 1; iteration time: 16951 ms; samples/sec: 58.994; batches/sec: 0.059; score: 61.64040843607799;
110852 [main] INFO  o.d.o.l.PerformanceListener - ETL: 6976 ms; iteration 2; iteration time: 12753 ms; samples/sec: 78.413; batches/sec: 0.078; score: 131.75989663147402;
123493 [main] INFO  o.d.o.l.PerformanceListener - ETL: 7690 ms; iteration 3; iteration time: 12641 ms; samples/sec: 79.108; batches/sec: 0.079; score: 139.66879515652315;

**gpu machine, cpu mode:**
38761 [main] INFO  o.d.o.l.PerformanceListener - ETL: 751 ms; iteration 0; iteration time: 3375 ms; samples/sec: 296.296; batches/sec: 0.296; score: 17.973563904870705;
44856 [main] INFO  o.d.o.l.PerformanceListener - ETL: 674 ms; iteration 1; iteration time: 6094 ms; samples/sec: 164.096; batches/sec: 0.164; score: 60.711884561402954;
49045 [main] INFO  o.d.o.l.PerformanceListener - ETL: 465 ms; iteration 2; iteration time: 4189 ms; samples/sec: 238.720; batches/sec: 0.239; score: 129.93905362345566;
52998 [main] INFO  o.d.o.l.PerformanceListener - ETL: 670 ms; iteration 3; iteration time: 3953 ms; samples/sec: 252.972; batches/sec: 0.253; score: 151.48013304498966;

**cpu machine (laptop, cpu only)**
INFO  09:16:54.501 [main           ] [PerformanceListener      ] ETL: 703 ms; iteration 0; iteration time: 2492 ms; samples/sec: 401.284; batches/sec: 0.401; score: 18.235179144737707;
INFO  09:16:56.990 [main           ] [PerformanceListener      ] ETL: 461 ms; iteration 1; iteration time: 2489 ms; samples/sec: 401.768; batches/sec: 0.402; score: 61.1129676358959;
INFO  09:16:58.967 [main           ] [PerformanceListener      ] ETL: 420 ms; iteration 2; iteration time: 1977 ms; samples/sec: 505.817; batches/sec: 0.506; score: 136.98393463715627;
INFO  09:17:00.830 [main           ] [PerformanceListener      ] ETL: 425 ms; iteration 3; iteration time: 1863 ms; samples/sec: 536.769; batches/sec: 0.537; score: 146.0492742728752;

gpu machine and cpu machine perform similarly when running in cpu mode, of course the gpu machine is slower as it has half the cores.
however when run in gpu mode performance is dismal.

can you suggest setup we could try? network configuration is similar to what is described in https://gist.github.com/fmorbini/936a7f1c9b0aa905e3ecf71dc096cb30

**Comment by raver119** [comment]
ETL: 6976 ms

That's your problem. What's your ETL pipeline?

**Comment by fmorbinitc** [comment]
what is ETL? it does not change between the cpu and gpu runs. only thing i change is remove the java path to cuda and remove the nd4j-cuda-* jars from the classpath.

**Comment by raver119** [comment]
ETL is how your data is converted to datasets. Your iterators you're using to feed nn training is your ETL.

So, what's your ETL?

**Comment by fmorbinitc** [comment]
even if etl would be 0 the iteration time would still be 5 times greater in gpu mode. sorry need to run now, will post more details on how we feed the data in a few hours. but again there is no difference on how we process the data in the 2 modes.

**Comment by raver119** [comment]
No man, it's not that simple. Your hardware isn't endless resource, so in all cases your ETL affects performance of training.

That's why I'm asking - what's your ETL?  You're doing something heavy there, and I need details.

**Comment by fmorbinitc** [comment]
we have a dataset iterator implementing MultiDataSetIterator that provides the data for each batch.
inside the method next we build a MultiDataSet we pass 8 INDArrays and build it like this:
super(new INDArray[]{vectorFeatures,ontologyFeatures},
        new INDArray[]{lblsIntent,lblsNE},
        new INDArray[]{vectorFeaturesMask,ontologyFeaturesMask},
        new INDArray[]{lblsMaskIntent,lblsMaskNE});

vectorFeatures is a 3d array, the 3 dimensions are:
1 the number of utterances in the batch
2 the number of features for each token (embedding dimentions)
3 the maximum number of tokens in an utterance

ontologyFeatures is another 3d array with a different dimension 2 (different features)

then we have the 2 labels and the masks (binary) to indicate when we don't care about the prediction.

the computationally intensive ones are possibly the 2 feature arrays. the first is built using a lookup into a word embeddings index using a map to go from string to long and using the long as a byte offset into a filechannel to read the embeddings.
For the ontology features we use a set of fsms for each feature and see which fsms the input token satisfies.

help me understand, why what we do for etl becomes a problem when using the gpu? you need extra resources to use the gpu and so the cpu becomes a bottleneck?

**Comment by fmorbinitc** [comment]
will try that, thanks

**Comment by raver119** [comment]
Awesome. If i'm right - your iteration time will stabilize. If I'm not - we'll see some new symptoms probably.

In any way - please keep us posted.

**Comment by fmorbinitc** [comment]
@raver119 sorry, can you please point me to the example you have in mind?
what we do is similar to this example: https://github.com/deeplearning4j/dl4j-examples/blob/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/recurrent/seq2seq/CustomSequenceIterator.java

**Comment by fmorbinitc** [comment]
@raver119 an update:
i've tried the .save and .load methods on MultiDataSet object to precompute all features and save them before training starts. During training the only thing the iterator does is the load method.

with a batch size of 10K (10 times larger than above) here is what i get. It's faster but still significantly slower than in cpu mode.

I store the files in /dev/shm that is a tmpfs (should be ram but i'm not sure) on the aws machine.

After a bit it stops due to some out of mem, not sure why that happens either.

138021 [main] INFO  o.d.o.l.PerformanceListener - ETL: 12169 ms; iteration 0; iteration time: 80133 ms; samples/sec: 124.793; batches/sec: 0.012; score: 18.242573386345715;
186093 [main] INFO  o.d.o.l.PerformanceListener - ETL: 11320 ms; iteration 1; iteration time: 48070 ms; samples/sec: 208.030; batches/sec: 0.021; score: 18.235388765830145;
218301 [main] INFO  o.d.o.l.PerformanceListener - ETL: 9700 ms; iteration 2; iteration time: 32207 ms; samples/sec: 310.492; batches/sec: 0.031; score: 18.009853910183555;
253897 [main] INFO  o.d.o.l.PerformanceListener - ETL: 10614 ms; iteration 3; iteration time: 35596 ms; samples/sec: 280.930; batches/sec: 0.028; score: 17.80819584351682;
Exception in thread "main" java.lang.RuntimeException: Failed to allocate 567000000 bytes from DEVICE [0] memory
	at org.nd4j.jita.memory.CudaMemoryManager.allocate(CudaMemoryManager.java:76)
	at org.nd4j.jita.workspace.CudaWorkspace.alloc(CudaWorkspace.java:213)
	at org.nd4j.jita.allocator.impl.AtomicAllocator.allocateMemory(AtomicAllocator.java:471)
	at org.nd4j.jita.allocator.impl.AtomicAllocator.allocateMemory(AtomicAllocator.java:416)
	at org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer.<init>(BaseCudaDataBuffer.java:255)
	at org.nd4j.linalg.jcublas.buffer.CudaFloatDataBuffer.<init>(CudaFloatDataBuffer.java:61)
	at org.nd4j.linalg.jcublas.buffer.factory.CudaDataBufferFactory.createFloat(CudaDataBufferFactory.java:331)
	at org.nd4j.linalg.factory.Nd4j.createBuffer(Nd4j.java:1500)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.<init>(BaseNDArray.java:289)
	at org.nd4j.linalg.jcublas.JCublasNDArray.<init>(JCublasNDArray.java:124)
	at org.nd4j.linalg.jcublas.JCublasNDArrayFactory.createUninitialized(JCublasNDArrayFactory.java:1614)
	at org.nd4j.linalg.factory.Nd4j.createUninitialized(Nd4j.java:4461)
	at org.nd4j.linalg.workspace.BaseWorkspaceMgr.createUninitialized(BaseWorkspaceMgr.java:296)
	at org.deeplearning4j.nn.layers.recurrent.CudnnLSTMHelper.backpropGradient(CudnnLSTMHelper.java:219)
	at org.deeplearning4j.nn.layers.recurrent.LSTMHelpers.backpropGradientHelper(LSTMHelpers.java:502)
	at org.deeplearning4j.nn.layers.recurrent.LSTM.backpropGradientHelper(LSTM.java:122)
	at org.deeplearning4j.nn.layers.recurrent.LSTM.backpropGradient(LSTM.java:93)
	at org.deeplearning4j.nn.layers.recurrent.BidirectionalLayer.backpropGradient(BidirectionalLayer.java:174)
	at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doBackward(LayerVertex.java:148)
	at org.deeplearning4j.nn.graph.ComputationGraph.calcBackpropGradients(ComputationGraph.java:2621)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1369)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1329)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:160)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:63)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.graph.ComputationGraph.fitHelper(ComputationGraph.java:1149)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1098)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1065)
	at com.tactile.argot.nlu.onenn.OneModel.trainNetwork(OneModel.java:478)
	at com.tactile.argot.nlu.onenn.OneModel.train(OneModel.java:222)
	at com.tactile.argot.nlu.composite.OntologyNLU.train(OntologyNLU.java:73)
	at com.tactile.argot.nlu.training.OntologyNLUModels.prepareV14Model(OntologyNLUModels.java:621)
	at com.tactile.argot.nlu.training.OntologyNLUModels.main(OntologyNLUModels.java:828)

I profiled the remote aws code and this is what i get:
![screen shot 2018-11-19 at 6 57 49 am](https://user-images.githubusercontent.com/22603687/48715190-48f88300-ebc9-11e8-8857-82149179c3e8.png)
The first row is the profiling of gpu code without precomputing, that is the code to calculate features is there (like at the beginning of this issue). In that case the majority of the time is spent in garbage collection, not sure why. i didn't profile the memory, i can do that if you think it's useful.
The second row is the code with precomputing, there you can see actual fit operations are getting some time but wait operations seem to be too heavy (hotspots).

not sure if the info is of any use to you, any advice? thanks

**Comment by raver119** [comment]
Show PerfListener output for smaller batch size: 256/512?

**Comment by raver119** [comment]
Also, please tell me size of single serialized dataset, i'll try to reproduce this issue locally

**Comment by fmorbinitc** [comment]
running the 256 experiment at the moment.
size of single dataset in the case of 10K batch size was around 33M bytes (size of serialized file on disk).

**Comment by fmorbinitc** [comment]
@raver119 here is the ETL info for 256 batch size case, first 20 iterations:

3395475 [main] INFO  o.d.o.l.PerformanceListener - ETL: 1405 ms; iteration 0; iteration time: 9843 ms; samples/sec: 26.008; batches/sec: 0.102; score: 18.15232703624806;
3399812 [main] INFO  o.d.o.l.PerformanceListener - ETL: 1321 ms; iteration 1; iteration time: 4337 ms; samples/sec: 59.027; batches/sec: 0.231; score: 17.983460623292984;
3402915 [main] INFO  o.d.o.l.PerformanceListener - ETL: 1034 ms; iteration 2; iteration time: 3103 ms; samples/sec: 82.501; batches/sec: 0.322; score: 18.995852123519533;
3407018 [main] INFO  o.d.o.l.PerformanceListener - ETL: 1085 ms; iteration 3; iteration time: 4103 ms; samples/sec: 62.393; batches/sec: 0.244; score: 19.05778558468622;
3411049 [main] INFO  o.d.o.l.PerformanceListener - ETL: 78 ms; iteration 4; iteration time: 4031 ms; samples/sec: 63.508; batches/sec: 0.248; score: 17.762127788516647;
3413107 [main] INFO  o.d.o.l.PerformanceListener - ETL: 65 ms; iteration 5; iteration time: 2058 ms; samples/sec: 124.393; batches/sec: 0.486; score: 18.31833226202926;
3416049 [main] INFO  o.d.o.l.PerformanceListener - ETL: 960 ms; iteration 6; iteration time: 2942 ms; samples/sec: 87.016; batches/sec: 0.340; score: 16.736513300771;
3419010 [main] INFO  o.d.o.l.PerformanceListener - ETL: 60 ms; iteration 7; iteration time: 2961 ms; samples/sec: 86.457; batches/sec: 0.338; score: 16.235420508595805;
3422088 [main] INFO  o.d.o.l.PerformanceListener - ETL: 71 ms; iteration 8; iteration time: 3078 ms; samples/sec: 83.171; batches/sec: 0.325; score: 16.521826175915162;
3425092 [main] INFO  o.d.o.l.PerformanceListener - ETL: 60 ms; iteration 9; iteration time: 3004 ms; samples/sec: 85.220; batches/sec: 0.333; score: 17.382304153042043;
3428064 [main] INFO  o.d.o.l.PerformanceListener - ETL: 972 ms; iteration 10; iteration time: 2972 ms; samples/sec: 86.137; batches/sec: 0.336; score: 16.33120707225146;
3431057 [main] INFO  o.d.o.l.PerformanceListener - ETL: 67 ms; iteration 11; iteration time: 2993 ms; samples/sec: 85.533; batches/sec: 0.334; score: 15.971231363672402;
3434041 [main] INFO  o.d.o.l.PerformanceListener - ETL: 73 ms; iteration 12; iteration time: 2984 ms; samples/sec: 85.791; batches/sec: 0.335; score: 16.769584922012893;
3437020 [main] INFO  o.d.o.l.PerformanceListener - ETL: 68 ms; iteration 13; iteration time: 2978 ms; samples/sec: 85.964; batches/sec: 0.336; score: 15.55782221419269;
3439063 [main] INFO  o.d.o.l.PerformanceListener - ETL: 62 ms; iteration 14; iteration time: 2043 ms; samples/sec: 125.306; batches/sec: 0.489; score: 15.986110289506193;
3442049 [main] INFO  o.d.o.l.PerformanceListener - ETL: 977 ms; iteration 15; iteration time: 2986 ms; samples/sec: 85.733; batches/sec: 0.335; score: 15.293667938423795;
3445046 [main] INFO  o.d.o.l.PerformanceListener - ETL: 970 ms; iteration 16; iteration time: 2997 ms; samples/sec: 85.419; batches/sec: 0.334; score: 15.342508540370146;
3449128 [main] INFO  o.d.o.l.PerformanceListener - ETL: 1016 ms; iteration 17; iteration time: 4082 ms; samples/sec: 62.714; batches/sec: 0.245; score: 13.99101006679934;
3452188 [main] INFO  o.d.o.l.PerformanceListener - ETL: 999 ms; iteration 18; iteration time: 3060 ms; samples/sec: 83.660; batches/sec: 0.327; score: 13.962916481107207;
3455170 [main] INFO  o.d.o.l.PerformanceListener - ETL: 72 ms; iteration 19; iteration time: 2982 ms; samples/sec: 85.848; batches/sec: 0.335; score: 13.61915423678386;

profiling is similar, 70% solver, 22% reading, 7% GC. hotspots are:
![screen shot 2018-11-19 at 8 32 24 am](https://user-images.githubusercontent.com/22603687/48720959-bf02e700-ebd5-11e8-8370-1b8f06cb7323.png)

**Comment by agibsonccc** [comment]
Closing — the CUDA backend has been completely rewritten with new memory management, DSP execution engine, CUDA graph replay, and Triton compilation. If you still experience GPU-related issues with the upcoming release, please open a new issue.
